src/main/java/com/salesforce/storm/spout/dynamic/DynamicSpout.java

```diff
@@ -399,7 +426,54 @@ public void removeVirtualSpout(final VirtualSpoutIdentifier virtualSpoutIdentifi
      * @return true when the spout exists, false when it does not.
      */
     public boolean hasVirtualSpout(final VirtualSpoutIdentifier spoutIdentifier) {
-        return getCoordinator().hasVirtualSpout(spoutIdentifier);
+        checkSpoutOpened();
+        return getSpoutMonitor().hasVirtualSpout(spoutIdentifier);
+    }
+
+    /**
+     * Create and start the SpoutMonitor instance.
+     * @param topologyConfig Topology configuration.
+     * @param topologyContext Topology context information.
+     * @param virtualSpoutMessageBus ThreadSafe message bus.
+     * @param metricsRecorder Metric reporter.
+     */
+    private void startSpoutMonitor(
+        final Map<String, Object> topologyConfig,
+        final TopologyContext topologyContext,
+        final VirtualSpoutMessageBus virtualSpoutMessageBus,
+        final MetricsRecorder metricsRecorder) {
+
+        // Prevent more than one instance from being created.
+        if (spoutMonitor != null) {
+            throw new IllegalStateException("Cannot create multiple spout monitor instances!");
+        }
+
+        // Build base name for threads: SpoutId:TaskIndex
+        final String threadContextName =
+            topologyContext.getThisComponentId() + ":" + topologyContext.getThisTaskIndex();
+
+        // Create instance.
+        spoutMonitor = new SpoutMonitor(
+            topologyConfig,
+            threadContextName,
+            virtualSpoutMessageBus,
+            metricsRecorder
+        );
+
+        // Create and name thread.
+        final Thread thread = new Thread(spoutMonitor, "Spout Monitor on " + threadContextName);
```

Comment:
Now if you have multiple spout instances on the same worker, the thread names will be unique like

"Spout Monitor on EventStreamSpout:0", "Spout Monitor on EventStreamSpout:1", "Spout Monitor on EventStreamSpout:2", etc..